data/tutorials/language/0lg_10_polymorphic_variants.md

```diff
@@ -0,0 +1,713 @@
+
+The following function was presented in the [Inferred Type Aliases](#Inferred-Type-Aliases) section.
+```ocaml
+# function `Avocado -> `Cilantro | plant -> plant;;
+- : ([> `Cilantro | `Avocado ] as 'a) -> 'a = <fun>
+```
+
+Because of the `plant -> plant` pattern, the types inferred as domain and codomain are the same. As `` `Avocado`` is accepted, it must be part of the domain; and as`` `Cilantro`` is returned, it must be part of the codomain. Both end up being part of the common type. However, `` `Avocado`` should not be part of the codomain, as this function can't possibly return this value. A finer type-checker would infer more precise types. Type-checking is an approximation and a trade-off, some valid programs are rejected, and some types are too coarse.
+
+### Performances
+
+TODO: Expand this section
+
+> There is one more downside: the runtime cost. A value `Pair (x, y)` occupies 3 words in memory, while a value `` `Pair (x, y)`` occupies 6 words.
+
+Guillaume Melquiond
+
+## When to Use Polymorphic Variant
+
+Only using polymorphic variants is over-engineering, it should be avoided. Polymorphic variants aren't an extension of simple variants. From the data perspective, polymorphic variants and simple variants are equivalent. They both are sum types in the algebraic data types sense, both with parametric polymorphism and recursion. Back quote isn't a difference that matters. The only meaningful difference is the algorithm. Therefore, deciding when to use polymorphic variants boils down to another question:
+> Are nominally or structurally type-checked variants needed?
+
+Answering this isn't significantly easier, but it helps to narrow what to consider. The key difference lies in the way pattern matching is type-checked. Polymorphic variant induces functions that intrinsically accept more data than simple variants. This is a consequence of the subtyping relation between polymorphic variant types.
+
+In a precautionary approach (inspired by [KISS](https://en.wikipedia.org/wiki/KISS_principle) or [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) design principles), using simple variants should be the default. However, they can feel too tight. Here are clues indicating this:
+* Many variant declarations look alike
+* Equivalent constructor duplicated in several variants
+* Variant declared for a single-purpose
+* Variant declaration feeling made up or artificial
+
+Having difficulties finding names for simple variants or constructors may be smell for one of the above. When such discomfort is felt, polymorphic variants may not be the solution, but they can be considered. Keep in mind they may ease some parts but at the expense of understandability, precision and performance. Hopefully, very soon, LLM bots will be able to refactor variants of some sort into the other. This will ease experimentation.
+
+To conclude, remember a simple variant that naturally encodes some data should remain a simple variant.
+
+## Conclusion
```

Comment:
Instead of a Conclusion i'd have a section that summarizes what was learned in this tutorial:

* We learned how to create new polymorphic variant tags
* We learned how to to constraint them 
* We learned how to reuse tags
* We learned how to use polymorphic variants for flexible error handling
* etc

Reply:
Yes, that's simpler.